### PR TITLE
Implement ar-file parsing in python

### PR DESCRIPTION
**Pull request by sbc100**
This takes a different approach to the problem of duplicate files
in ar archives.

This makes the emar wrapper unnecessary solving the problem at ar file
extraction time instead.

Fixes: #8670

### CODE REVIEW

File: emar.py [hunk 1]
```diff
@@ -6,98 +6,19 @@
...
 """Archive helper script
 
-This script acts as a frontend replacement for `ar`. See emcc.
-This is needed because, unlike a traditional linker, emscripten can't handle
-archive with duplicate member names.  This is because emscripten extracts
-archive to a temporary location and duplicate filenames will clobber each
-other in this case.
+This script is a simple wrapper around llvm-ar.  It used to have special
+handling for duplicate basenames in order to allow bitcode linking prcoess to
+read such files.  This is now hanled by using tools/arfile.py to read archives.
```
Comment (kripken) [inline emar.py]: typo in `handled`

File: tools/shared.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 from __future__ import print_function
 
 from distutils.spawn import find_executable
-from subprocess import PIPE, STDOUT
+from subprocess import PIPE, STDOUT # noqa
```
Comment (kripken) [inline tools/shared.py]: why is this needed?

File: tools/arfile.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
+# Copyright 2019 The Emscripten Authors.  All rights reserved.
+# Emscripten is available under two separate licenses, the MIT license and the
+# University of Illinois/NCSA Open Source License.  Both these licenses can be
+# found in the LICENSE file.
+
+"""Utility functions for parsing 'ar' files.
+
+This is needed in emsripten because command line tools such as llvm-ar are not
```
Comment (kripken) [inline tools/arfile.py]: typo in `emscripten`

File: tools/arfile.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+able to deal with archives containing many files with the same name.  Despite
+this linker are expected to handle this case and emscripten needs to emulate
```
Comment (kripken) [inline tools/arfile.py]: I'd remove `Despite this` (or add a comma perhaps).

File: tools/shared.py (hunk 2 above)
Comment (sbc100) [reply tools/shared.py]: There are no more direct uses of STDOUT in this file.. so this is an unused import.. but other modules expect to be able to import shared.STDOUT.

**Comment by kripken** [comment]
@sbc100 This broke the emscripten-releases Mac bot. That bot was broken on tests before, but this broke the build step.

autoroll: https://chromium-review.googlesource.com/c/emscripten-releases/+/1640373
failing build: https://ci.chromium.org/p/emscripten-releases/builders/ci/mac/b8911649373078606944
build from before for comparison (where the build step passes, but test step later fails): https://ci.chromium.org/p/emscripten-releases/builders/ci/mac/b8911885373953154080
